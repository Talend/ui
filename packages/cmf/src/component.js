import invariant from 'invariant';
import actionCreator from './actionCreator';
import expression from './expression';
import registry from './registry';
import CONST from './constant';

/**
 * All stuff related to the routing in CMF
 * @module react-cmf/lib/component
 * @see module:react-cmf/lib/component
 */

/**
 * return a component from the registry
 * @param  {string} id the component id you want
 * @param  {object} context optional context to get the registry
 * @return {function} the react component
 */
function get(id, context) {
	const component = registry.getFromRegistry(`${CONST.REGISTRY_COMPONENT_PREFIX}:${id}`, context);
	if (!component) {
		throw new Error(`component not found in the registry: ${id}`);
	}
	return component;
}

/**
 * register a component for the router configuration
 * @param  {string} id the component id you want to register
 * @param  {any} component the component you want to register
 * @param  {object} context optional context to get the registry
 */
function register(id, component, context) {
	if (!component) {
		invariant(process.env.NODE_ENV !== 'production', 'You can register undefined as a component');
		return;
	}
	registry.addToRegistry(`${CONST.REGISTRY_COMPONENT_PREFIX}:${id}`, component, context);
	if (component.actions) {
		Object.keys(component.actions).forEach(key => {
			actionCreator.register(key, component.actions[key], context);
		});
	}
	if (component.expressions) {
		Object.keys(component.expressions).forEach(key => {
			expression.register(key, component.expressions[key], context);
		});
	}
}

const registerMany = registry.getRegisterMany(register);

function has(id, context) {
	return (
		registry.getFromRegistry(`${CONST.REGISTRY_COMPONENT_PREFIX}:${id}`, context) !== undefined
	);
}

export default {
	get,
	has,
	register,
	registerMany,
};
