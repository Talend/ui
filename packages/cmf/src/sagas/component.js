import invariant from 'invariant';
import { fork, cancel, take, takeEvery } from 'redux-saga/effects';
import curry from 'lodash/curry';
import CONST from '../constant';
import registry from '../registry';

/**
 * This function register a saga in the cmf registry
 * @param {string} id the saga id you want
 * @param {generator} saga the saga generator
 * @param {object} context optional context to get the registry
 */
export function register(id, saga, context) {
	registry.addToRegistry(`SAGA:${id}`, saga, context);
}

/**
 * This function allow to get a saga from the registry
 * @param {string} id the saga id you want
 * @param {object} context optional context to get the registry
 */
export function get(id, context) {
	return registry.getFromRegistry(`SAGA:${id}`, context);
}

export const registerMany = registry.getRegisterMany(register);

export const isActionCancelable = curry(
	(startAction, action) =>
		action.type === `${CONST.WILL_UNMOUNT_SAGA_STOP}_${startAction.saga}` &&
		startAction.event.componentId === action.event.componentId,
);

export function* onSagaStart(action) {
	const saga = get(action.saga);
	if (!saga) {
		invariant(
			process.env.NODE_ENV === 'production',
			`You cannot register undefined as saga for id "${action.saga}"`,
		);
	} else {
		const task = yield fork(saga, action.props);
		yield take(isActionCancelable(action));
		yield cancel(task);
	}
}

export function* handle() {
	yield takeEvery(CONST.DID_MOUNT_SAGA_START, onSagaStart);
	yield take('DO_NOT_QUIT');
}
