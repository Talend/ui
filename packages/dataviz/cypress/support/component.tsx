/* eslint-disable @talend/import-depth */
// ***********************************************************
// This example support/component.ts is processed and
// loaded automatically before your test files.
//
// This is a great place to put global configuration and
// behavior that modifies Cypress.
//
// You can change the location of this file or turn off
// automatically serving support files with the
// 'supportFile' configuration option.
//
// You can read more here:
// https://on.cypress.io/configuration
// ***********************************************************

import './commands';

import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';
import { namespaces as designSystemNamespaces } from '@talend/locales-design-system/namespaces';
import { locales as designSystemLocales } from '@talend/locales-design-system/locales';

i18n.use(initReactI18next).init({
	debug: false,
	defaultNS: designSystemNamespaces[0],
	fallbackLng: 'en',
	fallbackNS: designSystemNamespaces,
	interpolation: {
		escapeValue: false,
	},
	ns: designSystemNamespaces,
	resources: Object.keys(designSystemLocales).reduce(
		(resources, language) => ({
			...resources,
			[language]: {
				...designSystemLocales[language],
			},
		}),
		{},
	),
	wait: true,
});
