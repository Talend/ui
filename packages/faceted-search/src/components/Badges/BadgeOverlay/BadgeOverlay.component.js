import React, { useState } from 'react';
import PropTypes from 'prop-types';
import Overlay from 'react-bootstrap/lib/Overlay';
import Popover from 'react-bootstrap/lib/Popover';
import Button from 'react-bootstrap/lib/Button';
import Icon from '@talend/react-components/lib/Icon';
import TooltipTrigger from '@talend/react-components/lib/TooltipTrigger';

const getChildren = (children, setOverlayOpened) => {
	if (typeof children === 'function') {
		return children(setOverlayOpened);
	}
	return children;
};

const getLabel = labels => {
	if (Array.isArray(labels)) {
		return labels.map(label => <span key={label}>{label}</span>);
	}
	return <span>{labels}</span>;
};

/**
 * The badge overlay can be use in two different ways.
 * An automatic one where you can only set an initial state.
 * Or a manual one where you have to provide an opened and a onChange function.
 * On hide can be used in both, returning the overlay setter,
 * which can be helpful if you are using the automatic way.
 */
const BadgeOverlay = ({
	children,
	className,
	iconName,
	initialOpened = false,
	id,
	label,
	onChange,
	onHide,
	opened = false,
	placement = 'bottom',
	rootClose = true,
}) => {
	const [overlayOpened, setOverlayOpened] = useState(initialOpened);
	const [buttonRef, setButtonRef] = useState(null);

	const changeOpened = event => {
		if (onChange) {
			onChange(event, !opened);
		} else {
			setOverlayOpened(!overlayOpened);
		}
	};

	const onHideOverlay = event => {
		if (onHide) {
			onHide(event, setOverlayOpened);
		} else {
			setOverlayOpened(false);
		}
	};
	const currentOpened = opened || overlayOpened;

	const button = (
		<Button
			id={`${id}-action-overlay`}
			bsStyle="link"
			aria-label={label}
			type="button"
			ref={target => setButtonRef(target)}
			onClick={changeOpened}
		>
			{iconName ? <Icon name={`talend-${iconName}`} key="icon" /> : getLabel(label)}
		</Button>
	);

	const buttonToRender = iconName ? (
		<TooltipTrigger label={label} tooltipPlacement="top">
			{button}
		</TooltipTrigger>
	) : (
		button
	);

	return (
		<div className={className}>
			{buttonToRender}

			<Overlay
				id={`${id}-overlay`}
				onHide={onHideOverlay}
				placement={placement}
				rootClose={rootClose}
				show={currentOpened}
				target={buttonRef}
			>
				<Popover id={`${id}-popover`}>{getChildren(children, setOverlayOpened)}</Popover>
			</Overlay>
		</div>
	);
};

BadgeOverlay.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.arrayOf(PropTypes.element),
		PropTypes.func,
	]).isRequired,
	className: PropTypes.string,
	iconName: PropTypes.string,
	initialOpened: PropTypes.bool,
	id: PropTypes.string.isRequired,
	label: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.string), PropTypes.string]),
	onChange: PropTypes.func,
	onHide: PropTypes.func,
	opened: PropTypes.bool,
	placement: PropTypes.string,
	rootClose: PropTypes.bool,
};

// eslint-disable-next-line import/prefer-default-export
export { BadgeOverlay };
