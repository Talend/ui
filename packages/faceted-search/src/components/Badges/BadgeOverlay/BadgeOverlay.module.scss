@use '~@talend/bootstrap-theme/src/theme/guidelines' as *;

.tc-badge-format-value {
	display: inline-flex;

	svg {
		margin: 0;
	}
}
