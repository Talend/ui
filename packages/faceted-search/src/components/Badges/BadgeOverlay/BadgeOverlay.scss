.tc-badge-link-plus {
	color: $lochmara;
	border: 1px solid rgba($lochmara, 0.25);
	text-transform: none;
	padding: 0 0.8rem;
	min-height: auto;
	height: 2.4rem;
	line-height: normal;
	border-radius: 2.5rem;
	font-weight: $font-weight-semi-bold;
	letter-spacing: normal;

	&:hover,
	&:focus {
		color: $lochmara;
		border-color: rgba($lochmara, 1);
		text-decoration: none;
	}

	.tc-badge-link-plus-icon {
		width: 1.2rem;
		height: 1.2rem;
		margin: 0;
	}
}
