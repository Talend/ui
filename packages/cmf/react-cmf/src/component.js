import invariant from 'invariant';
import actionCreator from './actionCreator';
import expression from './expression';
import sagas from './sagas';
import registry from './registry';
import CONST from './constant';

/**
 * All stuff related to the routing in CMF
 * @module react-cmf/lib/component
 * @see module:react-cmf/lib/component
 */

/**
 * return a component from the registry
 * @param  {string} id the component id you want
 * @param  {object} context optional context to get the registry
 * @return {function} the react component
 */
function get(id, context) {
	const component = registry.getFromRegistry(`${CONST.REGISTRY_COMPONENT_PREFIX}:${id}`, context);
	if (!component) {
		throw new Error(`component not found in the registry: ${id}`);
	}
	return component;
}

/**
 * register a component to let CMF be able to Inject it.
 * @param  {string} id the component id you want to register
 * @param  {any} component the component you want to register
 * @param  {object} context optional context to get the registry
 */
function register(id, component, context) {
	if (!component) {
		invariant(
			process.env.NODE_ENV === 'production',
			'You cannot register undefined as a component for id "%s"',
			id,
		);
		return;
	}
	registry.addToRegistry(`${CONST.REGISTRY_COMPONENT_PREFIX}:${id}`, component, context);
	if (component.actions) {
		actionCreator.registerMany(component.actions, context);
	}
	if (component.expressions) {
		expression.registerMany(component.expressions, context);
	}
	if (component.sagas) {
		sagas.registerMany(component.sagas, context);
	}
}

const registerMany = registry.getRegisterMany(register);

function has(id, context) {
	return (
		registry.getFromRegistry(`${CONST.REGISTRY_COMPONENT_PREFIX}:${id}`, context) !== undefined
	);
}

export default {
	get,
	has,
	register,
	registerMany,
};
