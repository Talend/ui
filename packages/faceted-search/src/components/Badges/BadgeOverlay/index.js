import { BadgeOverlay } from './BadgeOverlay.component';

export { BadgeOverlay };
