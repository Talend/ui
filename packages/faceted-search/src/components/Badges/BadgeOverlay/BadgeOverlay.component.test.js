/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';

import { BadgeOverlay } from './BadgeOverlay.component';

describe('BadgeOverlay', () => {
	it('should render the html output in the default state', () => {
		// Given
		const props = {
			id: 'my-id',
			label: 'my label',
		};
		// When
		const wrapper = mount(<BadgeOverlay {...props}>children</BadgeOverlay>);
		// Then
		expect(
			wrapper
				.find('Overlay')
				.at(0)
				.prop('show'),
		).toBe(false);
		expect(wrapper.html()).toMatchSnapshot();
	});
	it('should render the html output with children as function', () => {
		// Given
		const childrenAsFunc = () => <div id="my-children">hello world</div>;
		const props = {
			id: 'my-id',
			label: 'my label',
		};
		// When
		const wrapper = mount(<BadgeOverlay {...props}>{() => childrenAsFunc()}</BadgeOverlay>);
		// Then
		expect(wrapper.html()).toMatchSnapshot();
	});
	it('should change the show props overlay', () => {
		// Given
		const props = {
			id: 'my-id',
		};
		// When
		const wrapper = mount(<BadgeOverlay {...props} />);
		expect(
			wrapper
				.find('Overlay')
				.at(0)
				.prop('show'),
		).toBe(false);
		act(() => wrapper.find('button').simulate('click'));
		wrapper.update();
		// Then
		expect(
			wrapper
				.find('Overlay')
				.at(0)
				.prop('show'),
		).toBe(true);
	});
	it('should trigger a callback when button clicked', () => {
		// Given
		const onChange = jest.fn();
		const props = {
			id: 'my-id',
			onChange,
		};
		// When
		const wrapper = mount(<BadgeOverlay {...props} />);
		expect(
			wrapper
				.find('Overlay')
				.at(0)
				.prop('show'),
		).toBe(false);
		wrapper.find('button').simulate('click');
		// Then
		expect(onChange.mock.calls.length).toBe(1);
		expect(onChange.mock.calls[0][1]).toBe(true);
	});
	it('should render with the overlay showed', () => {
		// Given
		const props = {
			id: 'my-id',
			initialOpened: true,
		};
		// When
		const wrapper = mount(<BadgeOverlay {...props} />);
		// Then
		expect(
			wrapper
				.find('Overlay')
				.at(0)
				.prop('show'),
		).toBe(true);
	});
});
