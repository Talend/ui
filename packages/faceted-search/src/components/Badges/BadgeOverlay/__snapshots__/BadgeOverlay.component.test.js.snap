// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BadgeOverlay should render the html output in the default state 1`] = `
<div>
  <button id="my-id-action-overlay"
          aria-label="my label"
          type="button"
          class="btn btn-link"
  >
    <span>
      my label
    </span>
  </button>
</div>
`;

exports[`BadgeOverlay should render the html output with children as function 1`] = `
<div>
  <button id="my-id-action-overlay"
          aria-label="my label"
          type="button"
          class="btn btn-link"
  >
    <span>
      my label
    </span>
  </button>
</div>
`;
